microcosm_flask/swagger/parameters/base.py

```diff
@@ -0,0 +1,142 @@
+from abc import ABC, abstractmethod
+from pkg_resources import iter_entry_points
+from typing import Any, Iterable, Mapping, Optional, Sequence, Tuple
+
+from marshmallow.fields import Field, List, Nested
+
+from microcosm_flask.naming import name_for
+from microcosm_flask.swagger.naming import type_name
+
+
+ENTRY_POINT = "microcosm_flask.swagger.parameters"
+
+
+class ParameterBuilder(ABC):
+    """
+    Plugin-aware swagger parameter builder.
+
+    Discovers builder subclasses via the `microcosm_flask.swagger.parameters` entry point
+    and delegates to the first compatible implementation.
+
+    """
+    def __init__(self):
+        self.parsers = {
+            "default": self.parse_default,
+            "description": self.parse_description,
+            "enum": self.parse_enum_values,
+            "format": self.parse_format,
+            "$ref": self.parse_ref,
+            "type": self.parse_type,
+            "items": self.parse_items,
+        }
+
+    @classmethod
+    def build(cls, field: Field) -> Mapping:
+        """
+        Build a swagger parameter from a marshmallow field.
+
+        """
+        builders = [
+            entry_point.load()()
+            for entry_point in iter_entry_points(ENTRY_POINT)
+        ] + [
+            # put default last
+            cls.default_builder_type()(),
+        ]
+
+        builder = next(
+            builder
+            for builder in builders
+            if builder.supports_field(field)
+        )
+
+        return builder.build_parameter(field)
+
+    @classmethod
+    def default_builder_type(cls):
+        """
+        Define the default builder type.
+
+        """
+        # avoid circular dependencies
```

Comment:
I don't love this, but it turns out that some kinds of builders (e.g. lists) need the ability to reference the top-level `build` function directly. This means we either need to pass the builder (or a `graph`) around of ensure that every builder can call itself to `build` (via the above `@classmethod`). I went with the latter due to none of this code using the `graph` yet. This comes with the problem that we can't import the default implementation in the usual way because the base class has to finish loading first.

Reply:
It would, but it pretty much means introducing `graph` because there are recursive calls to this function so we need a back-reference.

This is probably the right thing to do.

Reply:
I don't see why we can't just split this into two classes, one which is a base class for all builders, and the other which is basically a registry for builders that selects which one to use